Web Inspector: objects can have multiple private fields with the same name

dcrousso opened the pull request:
#### eba01be516e6823677fedb7e3676d3e2cef5914e
<pre>
Web Inspector: objects can have multiple private fields with the same name
<a href="https://bugs.webkit.org/show_bug.cgi?id=256319">https://bugs.webkit.org/show_bug.cgi?id=256319</a>

Reviewed by Patrick Angle.

We should show all private fields for an object, not just the most recently defined one for each name.

Developers will be able to identify the value for `this.#foo` by looking for the last `#foo` in the list, as private fields must be declared as part of the `class` definition, meaning they will be created in that order.

* Source/JavaScriptCore/inspector/JSInjectedScriptHost.h:
* Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
(Inspector::JSInjectedScriptHost::getOwnPrivatePropertySymbols): Added.
(Inspector::JSInjectedScriptHost::getOwnPrivatePropertyDescriptors): Deleted.
* Source/JavaScriptCore/inspector/JSInjectedScriptHostPrototype.cpp:
(Inspector::JSInjectedScriptHostPrototype::finishCreation):
(Inspector::jsInjectedScriptHostPrototypeFunctionGetOwnPrivatePropertySymbols): Added.
(Inspector::jsInjectedScriptHostPrototypeFunctionGetOwnPrivatePropertyDescriptors): Deleted.
Instead of returning an object keyed by the stringified private symbol, mimic `Object.getOwnPropertySymbols` to just return the list of all private field symbols.
We can then use the existing code that calls `Object.getOwnPropertyDescriptor` (instead of doing that manually in C++).

* Source/JavaScriptCore/inspector/InjectedScriptSource.js:
(InjectedScript.prototype._forEachPropertyDescriptor.processProperty):
(InjectedScript.prototype._forEachPropertyDescriptor):
Since we&apos;re now using `Symbol`, differentiate private fields from regular (public) `Symbol` properties.

* LayoutTests/inspector/runtime/resources/property-descriptor-utilities.js:
* LayoutTests/inspector/runtime/getDisplayableProperties.html:
* LayoutTests/inspector/runtime/getDisplayableProperties-expected.txt:
* LayoutTests/inspector/runtime/getProperties.html:
* LayoutTests/inspector/runtime/getProperties-expected.txt:

Canonical link: <a href="https://commits.webkit.org/267483@main">https://commits.webkit.org/267483@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c27d580f920821b024f0e3f81b9f3b6d3f5f243e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14755 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15060 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15412 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16502 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13906 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17581 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15160 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16550 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14935 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15433 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12529 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17235 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12717 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13312 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20302 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/12633 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13791 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13479 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16716 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14024 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14021 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11860 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/14955 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13314 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/13168 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3825 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4001 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17651 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15185 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13865 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3635 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
```diff
@@ -315,7 +316,7 @@ JSValue JSInjectedScriptHost::getOwnPrivatePropertyDescriptors(JSGlobalObject* g
             continue;
 
-        result->putDirect(vm, Identifier::fromString(vm, String(propertyName.impl()->isolatedCopy())), objectConstructorGetOwnPropertyDescriptor(globalObject, object, propertyName));
+        result->putDirectIndex(globalObject, index++, Symbol::create(vm, *static_cast<SymbolImpl*>(propertyName.impl())));
```
@WebKit/jsc-reviewers will this cause issues?  AFAICT `Symbol::finishCreation` will add the `SymbolImpl` to a global (per-`VM`) `symbolImplToSymbolMap`, with no place that it's ever removed.  should I add something special to prevent the addition, or is this OK?

kmiller68 replied:
I think it should be fine because symbolImplToSymbolMap is a `WeakGCMap` so it should self clear when the `Symbol` gets GC'd.

kmiller68 commented on LayoutTests/inspector/runtime/resources/property-descriptor-utilities.js:
```diff
@@ -27,6 +51,30 @@ class PrivateMembersTestClassParent {
+    get #instancePrivateGetterSetter() { child }
+    set #instancePrivateGetterSetter(x) { child }
+    static classPublicProperty = 'classPublicPropertyValue child';
+    static #classPrivateProperty = 'classPrivatePropertyValue child';
```
Should this appear somewhere in your output? I don't see it if so.

dcrousso replied:
[LayoutTests/inspector/runtime/getDisplayableProperties-expected.txt:152](https://github.com/WebKit/WebKit/pull/13448/files#diff-a4beb82f7378e7e9b9b285825d88387629c0b16f05ca36fd2a5580f2c70e3543R152)
[LayoutTests/inspector/runtime/getProperties-expected.txt:134](https://github.com/WebKit/WebKit/pull/13448/files#diff-f594c930ddabcad1999b2b74a9f82be7773b1a9c65f1035cfb88fec84882932eR134)

also the `'classPrivatePropertyValue parent'` would be visible if we tested `PrivateMembersTestClassParent`

dcrousso committed:
Web Inspector: objects can have multiple private fields with the same name
https://bugs.webkit.org/show_bug.cgi?id=256319

Reviewed by Patrick Angle.

We should show all private fields for an object, not just the most recently defined one for each name.

Developers will be able to identify the value for `this.#foo` by looking for the last `#foo` in the list, as private fields must be declared as part of the `class` definition, meaning they will be created in that order.

* Source/JavaScriptCore/inspector/JSInjectedScriptHost.h:
* Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
(Inspector::JSInjectedScriptHost::getOwnPrivatePropertySymbols): Added.
(Inspector::JSInjectedScriptHost::getOwnPrivatePropertyDescriptors): Deleted.
* Source/JavaScriptCore/inspector/JSInjectedScriptHostPrototype.cpp:
(Inspector::JSInjectedScriptHostPrototype::finishCreation):
(Inspector::jsInjectedScriptHostPrototypeFunctionGetOwnPrivatePropertySymbols): Added.
(Inspector::jsInjectedScriptHostPrototypeFunctionGetOwnPrivatePropertyDescriptors): Deleted.
Instead of returning an object keyed by the stringified private symbol, mimic `Object.getOwnPropertySymbols` to just return the list of all private field symbols.
We can then use the existing code that calls `Object.getOwnPropertyDescriptor` (instead of doing that manually in C++).

* Source/JavaScriptCore/inspector/InjectedScriptSource.js:
(InjectedScript.prototype._forEachPropertyDescriptor.processProperty):
(InjectedScript.prototype._forEachPropertyDescriptor):
Since we're now using `Symbol`, differentiate private fields from regular (public) `Symbol` properties.

* LayoutTests/inspector/runtime/resources/property-descriptor-utilities.js:
* LayoutTests/inspector/runtime/getDisplayableProperties.html:
* LayoutTests/inspector/runtime/getDisplayableProperties-expected.txt:
* LayoutTests/inspector/runtime/getProperties.html:
* LayoutTests/inspector/runtime/getProperties-expected.txt:

Canonical link: https://commits.webkit.org/267483@main
